### PR TITLE
IBP-2909 - Import Crosses Optimization

### PR DESCRIPTION
**Pull request by mderamos**
Kindly Review. Thanks!

### CODE REVIEW

File: src/main/java/com/efficio/fieldbook/web/common/service/impl/CrossingTemplateParser.java [hunk 1]
```diff
@@ -14,23 +16,23 @@
...
 import org.generationcp.commons.parsing.pojo.ImportedFactor;
+import org.generationcp.commons.parsing.pojo.ImportedGermplasmParent;
 import org.generationcp.commons.parsing.pojo.ImportedVariate;
 import org.generationcp.commons.spring.util.ContextUtil;
 import org.generationcp.commons.util.DateUtil;
-import org.generationcp.middleware.domain.gms.GermplasmListType;
 import org.generationcp.middleware.exceptions.MiddlewareQueryException;
 import org.generationcp.middleware.manager.api.StudyDataManager;
 import org.generationcp.middleware.pojos.Germplasm;
 import org.generationcp.middleware.pojos.ListDataProject;
```
Comment (darla-leafnode) [inline src/main/java/com/efficio/fieldbook/web/common/service/impl/CrossingTemplateParser.java]: Remove `ListDataProject` import

File: src/main/java/com/efficio/fieldbook/web/common/service/impl/CrossingTemplateParser.java [hunk 2]
```diff
@@ -149,37 +154,139 @@ private void parseObservationSheet(final String programUUID) throws FileParsingE
...
+	 */
+
+	void lookupCrossParents(final String femaleStudyName, final Set<Integer> femalePlotNos,
+		final Map<String, Set<Integer>> maleStudiesWithPlotNos, final Map<Integer, Triple<String, Integer, List<Integer>>> entryIdToCrossInfoMap,
+		final String programUUID) throws FileParsingException {
+
+		final Map<Integer, ImportedCrossParent> femalePlotMap =
+			this.getPlotToImportedCrossParentMapForStudy(femaleStudyName, femalePlotNos, programUUID);
+		// Create map of male studies to its plotToListDataProject lookup
+		// for each male study, lookup the associated ListDataProject of specified male plot #s
```
Comment (darla-leafnode) [inline src/main/java/com/efficio/fieldbook/web/common/service/impl/CrossingTemplateParser.java]: Update comments and variables using `ListDataProject`

File: src/main/java/com/efficio/fieldbook/web/common/service/impl/CrossingTemplateParser.java [hunk 3]
```diff
@@ -149,37 +154,139 @@ private void parseObservationSheet(final String programUUID) throws FileParsingE
...
+	 * Returns map of plot numbers to corresponding ImportedCrossParent record from specified plot numbers
+	 * of given study
+	 *
+	 * @param studyName
+	 * @param plotNos
+	 * @param programUUID
+	 * @return
+	 * @throws FileParsingException
+	 */
+	Map<Integer, ImportedCrossParent> getPlotToImportedCrossParentMapForStudy(final String studyName, final Set<Integer> plotNos,
```
Comment (darla-leafnode) [inline src/main/java/com/efficio/fieldbook/web/common/service/impl/CrossingTemplateParser.java]: There is no need to pass around `programUUID` as parameter through several methods from `parseWorkbook `method.
We can just use `this.contextUtil.getCurrentProgramUUID()` in the actual method where it's called which is `getPlotToImportedCrossParentMapForStudy`

File: src/main/java/com/efficio/fieldbook/web/common/service/impl/CrossingTemplateParser.java [hunk 4]
```diff
@@ -149,37 +153,136 @@ private void parseObservationSheet(final String programUUID) throws FileParsingE
...
-			importedCrosses.setCross(crossString);
+			germplasm.setGpid1(Integer.valueOf(cross.getFemaleGid()));
+			germplasm.setGpid2(cross.getMaleParents().get(0).getGid());
+			final String crossString = this.crossingService.getCross(germplasm, cross, "/");
+			cross.setCross(crossString);
+		}
+	}
 
-			this.importedCrossesList.addImportedCrosses(importedCrosses);
+	List<ImportedGermplasmParent> getMaleParents(
```
Comment (clarysabel) [inline src/main/java/com/efficio/fieldbook/web/common/service/impl/CrossingTemplateParser.java]: getMaleParents name was a bit confusing to me given that it is a cross attribute and in fact is loading the male parents. Probably we should use a more appropriated name @darla-leafnode

Comment (darla-leafnode) [reply src/main/java/com/efficio/fieldbook/web/common/service/impl/CrossingTemplateParser.java]: A cross here can have one or multiple male parents (polycross) as opposed to just having only one female parent at all time)

**Pull request merged**